csharp/ql/src/semmle/code/csharp/controlflow/Guards.qll

```diff
@@ -1454,6 +1630,59 @@ module Internal {
             forex(PreSsa::Definition u | u = def.getAnUltimateDefinition() | nonNullDef(u))
           ).getARead()
       }
+
+      private predicate adjacentReadPairSameVarUniquePredecessor(
+        AssignableRead read1, AssignableRead read2
+      ) {
+        PreSsa::adjacentReadPairSameVar(read1, read2) and
+        not exists(AssignableRead other |
+          PreSsa::adjacentReadPairSameVar(other, read2) and
+          other != read1 and
+          other != read2
+        )
+      }
+
+      cached
+      predicate emptyValue(Expr e) {
+        e.(ArrayCreation).getALengthArgument().getValue().toInt() = 0
+        or
+        exists(Expr mid | emptyValue(mid) |
+          mid = e.(AssignExpr).getRValue()
+          or
+          mid = e.(Cast).getExpr()
+        )
+        or
+        exists(PreSsa::Definition def | emptyDef(def) | firstReadSameVarUniquePredecesssor(def, e))
+        or
+        exists(MethodCall mc |
+          mc.getTarget().getSourceDeclaration() = any(SystemCollectionsGenericListClass c)
```

Comment:
We could handle other containers at this point, instead of just `List`. For example, we could use `IList<T>` or perhaps best of all, `ICollection<T>`.

Reply:
Yes, `ICollection` is better.